src/origin-experiments.js

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {bytesToString, bytesToUInt32, stringToBytes} from './utils/bytes';
+import {getSourceOrigin, parseUrl} from './url';
+import {parseJson} from './json';
+
+/** @const {number} */
+const LENGTH_BYTES = 4;
+
+/**
+ * Generates, signs and verifies origin experiments.
+ */
+export class OriginExperiments {
+  /**
+   * @param {!./service/crypto-impl.Crypto} crypto
+   */
+  constructor(crypto) {
+    /** @private {!./service/crypto-impl.Crypto} */
+    this.crypto_ = crypto;
+  }
+
+  /**
+   * Generates an RSA public/private key pair for signing and verifying.
+   * @return {!Promise}
+   */
+  generateKeys() {
+    const generationAlgo = Object.assign({
+      modulusLength: 2048,
+      publicExponent: Uint8Array.of(1, 0, 1),
```

Comment:
How'd you come up with this?

Reply:
The same code existed in the now-deleted `generate-testdata-experiments.js_`.